File: src/ShopNet/Gallery/Get.ts
import { ShopNet } from "../ShopNet";

/**
 * @author Aloento
 * @since 0.5.0
 * @version 0.1.0
 */
export abstract class GalleryGet extends ShopNet {
  /**
   * @author Aloento
   * @since 0.5.0
   * @version 0.2.1
   */
  public static Categories(): Promise<string[]> {
    return this.GetTimeCache("", "GalleryGetCategories", (x) => x.add(1, "m"));
  }

  /**
   * @author Aloento
   * @since 0.5.0
   * @version 0.2.2
   */
  public static async Products(category: string): Promise<number[]> {
    const nums = await this.GetTimeCache<number[]>(category, "GalleryGetProducts", (x) => x.add(1, "m"), category);
    return nums;
  }
}

File: SoarCraft.AwaiShop.Test/Hub/Gallery/Get.cs
namespace SoarCraft.AwaiShop.Test.Hub.Gallery;

using Microsoft.AspNetCore.SignalR.Client;
using Newtonsoft.Json;
using System.Diagnostics;

[TestClass]
[TestCategory("Hub.Gallery.Get")]
public class Get : ShopNet
{
    public static Task<string[]> GalleryGetCategories() =>
        Guest.InvokeAsync<string[]>(nameof(GalleryGetCategories));

    /**
     * <remarks>
     * @Author Aloento
     * @TargetVersion 0.2.0
     * </remarks>
     */
    [TestMethod]
    public async Task Categories()
    {
        var res = await GalleryGetCategories();
        Assert.IsTrue(res.Length > 0);
        Debug.WriteLine(JsonConvert.SerializeObject(res, Formatting.Indented));
    }

    public static Task<uint[]> GalleryGetProducts(string category) =>
        Guest.InvokeAsync<uint[]>(nameof(GalleryGetProducts), category);

    /**
     * <remarks>
     * @Author Aloento
     * @TargetVersion 0.2.0
     * </remarks>
     */
    [TestMethod]
    public async Task Products()
    {
        var categories = await GalleryGetCategories();
        Assert.IsTrue(categories.Length > 0);

        foreach (var category in categories)
        {
            var res = await GalleryGetProducts(category);
            Assert.IsTrue(res.Length > 0);
            Debug.WriteLine($"Category [{category}] \n" + JsonConvert.SerializeObject(res, Formatting.Indented));
        }
    }
}
